File: backend/src/utils/transcript.js
const fs = require("fs");

function convertTranscriptToText(data) {

    if(typeof data === "string"){
        data = JSON.parse(data);
    }

    function formatTime(ms) {
        const h = Math.floor(ms / 3600000);
        const m = Math.floor((ms % 3600000) / 60000);
        const s = Math.floor((ms % 60000) / 1000);
        const msPart = ms % 1000;

        return `${String(h).padStart(2,"0")}:${String(m).padStart(2,"0")}:${String(s).padStart(2,"0")}.${String(msPart).padStart(3,"0")}`;
    }

    function appendWord(sentence,text){
        if(!sentence) return text.trim();

        if(/^\s/.test(text))
            return sentence + text;

        if(/^[,.;!?:"')\]]/.test(text))
            return sentence + text;

        return sentence + " " + text;
    }

    function cleanText(text){
        return text
            .replace(/>>/g,"")
            .replace(/\[[^\]]*\]/gi,"")
            .replace(/[♪♫♬]/g,"")
            .trim();
    }

    function wordCount(text){
        return text.split(/\s+/).filter(Boolean).length;
    }

    function isEndSentence(text){
        return /[.!?]["')\]]*$/.test(text.trim());
    }

    function saveSentence(result,start,text){
        text = text.replace(/\s+/g," ").trim().replace(/[.,!?…]+$/g, "");;

        if(!text) return;

        result.push(
            `${formatTime(start)} ${text}`
        );
    }

    let result = [];
    let current = "";
    let startTime = null;
    let buffer = [];

    for(const event of data.events){

        if(!event.segs) continue;

        for(const seg of event.segs){

            if(!seg.utf8) continue;

            let text = cleanText(seg.utf8);

            if(!text) continue;

            let time =
                (event.tStartMs || 0)
                +
                (seg.tOffsetMs || 0);

            if(startTime === null)
                startTime = time;

            current = appendWord(current,text);

            if(isEndSentence(text)){

                if(buffer.length){
                    current =
                        buffer.join(" ")
                        + " "
                        + current;

                    buffer = [];
                }

                saveSentence(
                    result,
                    startTime,
                    current
                );

                current = "";
                startTime = null;
            }
        }
    }

    if(current.trim()){

        if(buffer.length){
            current =
                buffer.join(" ")
                + " "
                + current;
        }

        saveSentence(
            result,
            startTime || 0,
            current
        );
    }

    return result.join("\n");
}

module.exports = {
    convertTranscriptToText
};

File: backend/app/utils/transcript.py
import json
import re


def format_time(ms: int) -> str:

    h = ms // 3_600_000

    m = (ms % 3_600_000) // 60_000

    s = (ms % 60_000) // 1_000

    ms_part = ms % 1_000

    return (
        f"{h:02d}:"
        f"{m:02d}:"
        f"{s:02d}."
        f"{ms_part:03d}"
    )


def append_word(
    sentence: str,
    text: str
) -> str:

    if not sentence:
        return text.strip()

    if re.match(r"^\s", text):
        return sentence + text

    if re.match(
        r"^[,.;!?:\"')\]]",
        text
    ):
        return sentence + text

    return sentence + " " + text


def clean_text(text: str) -> str:

    text = re.sub(
        r">>",
        "",
        text
    )

    text = re.sub(
        r"\[[^\]]*\]",
        "",
        text,
        flags=re.IGNORECASE
    )

    text = re.sub(
        r"[♪♫♬]",
        "",
        text
    )

    return text.strip()


def is_end_sentence(text: str) -> bool:

    return bool(
        re.search(
            r'[.!?]["\')\]]*$',
            text.strip()
        )
    )


def save_sentence(
    result: list,
    start: int,
    text: str
):

    text = re.sub(
        r"\s+",
        " ",
        text
    ).strip()

    text = re.sub(
        r"[.,!?…]+$",
        "",
        text
    )

    if not text:
        return

    result.append(
        f"{format_time(start)} {text}"
    )


def convert_transcript_to_text(data) -> str:

    if isinstance(data, str):
        data = json.loads(data)

    result = []

    current = ""

    start_time = None

    buffer = []

    for event in data.get("events", []):

        segs = event.get("segs")

        if not segs:
            continue

        for seg in segs:

            if not seg.get("utf8"):
                continue

            text = clean_text(
                seg["utf8"]
            )

            if not text:
                continue

            time = (
                event.get("tStartMs", 0)
                +
                seg.get("tOffsetMs", 0)
            )

            if start_time is None:
                start_time = time

            current = append_word(
                current,
                text
            )

            if is_end_sentence(text):

                if buffer:

                    current = (
                        " ".join(buffer)
                        + " "
                        + current
                    )

                    buffer = []

                save_sentence(
                    result,
                    start_time,
                    current
                )

                current = ""

                start_time = None

    if current.strip():

        if buffer:

            current = (
                " ".join(buffer)
                + " "
                + current
            )

        save_sentence(
            result,
            start_time or 0,
            current
        )

    return "\n".join(result)